4599 restyle product listings

mkllnk opened the pull request:
#### What? Why?

Closes #4599

<!-- Explain why this change is needed and the solution you propose.
Provide context for others to understand it. -->

This pull request includes the following changes:

- No more red, grey, black colour palette.
- Photos now display on mobile
- Product title wraps (same as current version)
- Producer name truncates
- Weight/size wraps
- Price breakdown "pie" displays on mobile. *(Some of the pop-up is not visible on mobile, it's partly off-screen but that should be solved in #4600.)*
- Listing content layout change
- Removal of category icon
- Display product properties under description
- Display truncated product description, 'more' link to open product detail modal. *(The designs show a two-line description on small screens and a "more" button. I went for a simpler implementation that show the ellipsis after the first line. [Yuko gave her okay](https://github.com/openfoodfoundation/openfoodnetwork/issues/4599#issuecomment-617518941).)*
- Clicking on product name or image also opens product detail modal
- Removal of per-product sub-total display on mobile
- Categories design has less padding without affecting the categories design in the filter section.

![Screenshot from 2020-04-28 16-55-47](https://user-images.githubusercontent.com/3524483/80457786-39ab6680-8973-11ea-9436-ffa365a2682f.png)
![Screenshot from 2020-04-28 17-11-34](https://user-images.githubusercontent.com/3524483/80457904-61023380-8973-11ea-922c-5469f8765d33.png)


#### What should we test?
<!-- List which features should be tested and how. -->

Create a shopfront with:

* A short and a long producer name.
* Short and long product names.
* Short and long product descriptions.
* Some product properties.
* Products with and without bulk buy.

Then test the shopfront display in multiple browsers at multiple screen sizes. Resizing the window can be good to spot glitches between typical sizes.

Add products to you cart to make sure that it works as before.


#### Release notes
<!-- Write a line or two to be included in the release notes.
Everything is worth mentioning, because you did it for a reason. -->

Updated product list design

<!-- Please assign one category to your PR and delete the others. 
The categories are based on https://keepachangelog.com/en/1.0.0/. -->

Changelog Category: Changed

#### Dependencies
<!-- Does this PR depend on another one?
Add the link or remove this section. -->

This will be merged into a feature branch. It will need two more issues before going live:

- #4598
- #4600


#### Documentation updates
<!-- Are their any wiki pages that need updating after merging this PR?
List them here or remove this section. -->

Some screenshots may need replacing.

luisramos0 commented:
looks like there's too many commits on this PR... some rebase may be needed on the target branch?

mkllnk commented:
Ah, yes, I accidentally based this on master and didn't update the feature branch. Should be sorted now.

mkllnk commented:
Thanks for your OCD, @Matt-Yorkley! Good attention to detail.

I added another commit to align the prices. The graph button had a silly padding. I'll leave the alignment with the input field for later because that's going to be replaced.

Vertical align will be a pain with the current structure. I'm wondering if we need to replace the foundation styles here. They use `float: left` for the columns which means that the divs don't have a consistent height and we can't do a vertical align. Any idea about that, @Matt-Yorkley?

The other issue is that the left column can have up to three lines which is not in the design:
![Screenshot from 2020-05-04 10-29-34](https://user-images.githubusercontent.com/3524483/80929816-507a0f00-8df2-11ea-9bda-681cb8ca7485.png)

luisramos0 commented on app/assets/stylesheets/darkswarm/_shop-product-rows.css.scss:
```diff
@@ -9,44 +9,13 @@
-          background-color: #f9f9f9;
-
-          &:hover, &:focus, &:active {
-            background-color: $clr-brick-ultra-light;
```
this makes $clr-brick-ultra-light in the branding css file dead code, can be deleted as well.

luisramos0 commented on app/views/shop/products/_summary.html.haml:
```diff
@@ -1,21 +1,18 @@
-          %i.ofn-i_036-producers
           %span{"ng-bind" => "::enterprise.name"}
-  .small-2.medium-2.large-1.columns.text-center
-    .taxon-flag
```
this makes this line dead:
app/assets/stylesheets/darkswarm/shop.css.scss
   11: @import "shop-taxon-flag";
and the file
app/assets/stylesheets/darkswarm/_shop-taxon-flag.css.scss:

these can be deleted as well, right?

luisramos0 commented on app/assets/stylesheets/darkswarm/_shop-product-rows.css.scss:
```diff
@@ -140,16 +140,26 @@
             &:hover, &:focus, &:active {
-              color: $clr-brick;
+              color: $orange-600;
+	      text-decoration: underline;
```
there's an indentation problem here

luisramos0 commented on app/assets/javascripts/templates/shop_variant.html.haml:
```diff
@@ -6,20 +6,16 @@
       %em><
         \ {{::'bulk' | t}}
+  .small-3.medium-3.large-3.columns.variant-price
+    -# Now in a template in app/assets/javascripts/templates !
```
I think this comment can be deleted.
templates in app/assets/javascripts/templates are not a new thing now.

luisramos0 commented on app/assets/stylesheets/darkswarm/_shop-product-rows.css.scss:
```diff
@@ -142,6 +142,22 @@
             text-overflow: ellipsis;
 	  }
+
+          .product-properties {
```
there's an indentation problem here and then below in has-tip as well

mkllnk commented:
Thank you @luisramos0! Well spotted. I addressed all your comments. Most of them I put in new commits which I can squash at the end of the review. I fixed the spacing in the commits where I introduced them though, because there were multiple.

sauloperez commented on app/assets/stylesheets/darkswarm/_shop-popovers.css.scss:
```diff
@@ -5,18 +5,17 @@
     width: 98%;
-    background-color: white;
+    background-color: $grey-800;
+    color: white;
```
This is the same I spotted in https://github.com/openfoodfoundation/openfoodnetwork/pull/5263/files#r420699417.

mkllnk replied:
Okay, maybe that needs another pull request for all the places where we use `white`. But who is to say if we want to change them all together or only some of them? `$white` doesn't have a semantic meaning. I would leave it until we actually need to change it and have meaning for it.

sauloperez commented on app/assets/stylesheets/darkswarm/_shop-popovers.css.scss:
```diff
@@ -25,54 +24,27 @@
-      background-color: #0b8c61;
-    }
+  li {
+    border-bottom: .1em solid $grey-700;
```
isn't probably 1px what you want? I can't imagine a border scaling according to the font size. I would look broken, I guess.

mkllnk replied:
I usually prefer `em` over `px`. It doesn't look odd at all. Actually, browser make `1px` bigger than one pixel when zooming in so that it doesn't look odd.

No zoom:
![Screenshot from 2020-05-07 16-00-59](https://user-images.githubusercontent.com/3524483/81260048-8fbb9080-907c-11ea-991b-8d4e69bef30f.png)

With zoom:
![Screenshot from 2020-05-07 16-01-50](https://user-images.githubusercontent.com/3524483/81260061-96e29e80-907c-11ea-8b51-b32b602b7849.png)

I think it's good scale everything and keep relative proportions. When an element becomes bigger, it needs a bigger border as well.

sauloperez replied:
well, I'm not so sure. This is the sort of question we need to ask the designer.

mkllnk replied:
@yukoosawa Do you prefer border width defined in `px` or `em` or something else? Does it depend on the case or is there a general rule?

sauloperez commented on app/assets/stylesheets/darkswarm/_shop-popovers.css.scss:
```diff
@@ -25,54 +24,27 @@
+  li {
+    border-bottom: .1em solid $grey-700;
+    padding: 0 0.25rem;
+    margin-bottom: 2px;
```
IMO we should stick to either `px` or `rem` and I lean towards the latter so it scales with the font size.

mkllnk replied:
I agree. But I didn't change this line, except for indentation.

sauloperez commented on app/assets/stylesheets/darkswarm/_shop-product-rows.css.scss:
```diff
@@ -141,37 +83,53 @@
+          .product-description {
+            white-space: nowrap;
+            overflow: hidden;
+            text-overflow: ellipsis;
```
these three properties are duplicated in `.product-producer` so I think it's worth extracting this pattern and giving it a name.

mkllnk replied:
I usually start DRYing something when I use it for the third time. These properties are definitely right for the producer. But for the product description, it's only an approximation of the design created by Yuko. She wanted a "more" button at the end and display two lines on mobile. That is far more complex and needs Javascript. I avoided that complexity for now but maybe someone wants to change this logic in the future. And then the extraction would just be unnecessary complexity.

In the words of the following article, I would call this *accidental* duplication.
https://www.justinweiss.com/articles/i-dry-ed-up-my-code-and-now-its-hard-to-work-with-what-happened/

sauloperez commented on app/assets/stylesheets/darkswarm/_shop-product-thumb.css.scss:
```diff
@@ -1,80 +1,10 @@
-          display: none;
-          width: 0rem;
-          height: 0rem;
-        }
```
there's little left in this file after this change so perhaps we can reconsider its existence. Shouldn't it be part of _shop-product-rows.css.scss  ?

mkllnk replied:
Yes, it's small but also simple. :smile: 
I find _shop-product-rows.css.scss a bit complicated already. I would be happy either way but avoided the additional commit here.

sauloperez commented on app/assets/stylesheets/darkswarm/_shop-popovers.css.scss:
```diff
@@ -25,54 +24,27 @@
+  ul, li {
+    list-style: none;
+    margin: 0;
+    font-size: 0.875rem;
```
we need to come up with a set of defined font-size vars as we do with branding.css so we pick consistent values. This won't be easy but we got to start agreeing on a convention as we move along.

I did a bit of gripping and so far we have around 72 different values for `font-size` across the app :boom: this can't end well. But this `0.875rem;` seems to be a pretty standard-ish font size.

I suggest we start adding a var for the font-sizes you are using here in https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/assets/stylesheets/darkswarm/typography.css.scss. It seems the team came up with a similar idea with `.text-small` there. This `0.875rem;`.

Then, after this PR, we can think of how to reuse these vars and not dying in the effort. Thoughts?

Matt-Yorkley replied:
Agreed! We should figure out `rem` and `em` guidelines. I think `em` is generally good for font-sizes but maybe `rem` for padding and margin..?

mkllnk replied:
I agree that it's a mess and it would be good to standardise these things. I would prefer to do that in another PR though because this one has 25 commits already.

mkllnk commented:
@sauloperez Are you okay for this pull request to go to test ready? I would be good to have this in the feature branch to base other work on it.

daniellemoorhead commented:
>@sauloperez Are you okay for this pull request to go to test ready? I would be good to have this in the feature branch to base other work on it.

Please please please make it so! We've been waiting so long to have the soviet brutalist architecture vibe removed from our product list view. Release the food images!

![](https://media.giphy.com/media/VLoN2iW8ii3wA/giphy.gif)

daniellemoorhead commented:
I've started to test this, [testing notes are here](https://docs.google.com/document/d/1WZZSuMCLHiZ_UY_nLJZ8MI-HGadTQmVty7zXeS7H6VE/edit?usp=sharing).

@mkllnk there are a few things that require fixing, would be great if you could do those before we have @yukoosawa look at this from a designer perspective.

I'm popping this back to in dev for you to do that first pass. Ta!

mkllnk commented:
Great testing, @daniellemoorhead!

> 1. Product name displays 2nd time above the unit measure

It's not the product name that shows again, it's the variant name. And the variant name is often the same as the product name but not always. The design shows this:

![Screenshot from 2020-05-13 12-29-51](https://user-images.githubusercontent.com/3524483/81764723-88c9cd80-9515-11ea-972b-e85faf3e8cd2.png)

I'm not sure how to resolve this. We show the product summary (image, name, description, supplier) and then we list all variants. Should I hide the variant name if it's the same as the product name? What if only one variant has the same name and the others are different? Treating them differently would look odd.

What are your thoughts, @daniellemoorhead?

> 2. Product description only shows 1 line of copy, in the original designs it shows 2 lines.

I know, that's because HTML can't display the ... after two lines of text. It only works for one. There are more complicated Javascript solutions but I was hoping to avoid those. How important is this second line?

The rest of the points look easy to fix to me:

- [ ] Listing image should bump up against the left hand side edge of the browser window, but at the moment there is a significant amount of padding.
- [ ] Padding on right hand side looks to be double what it should be compared to the original designs.
- [ ] Spacing between description / properties and price is not as wide as in the original designs.
- [ ] The properties have a hover state (darker grey border) and are clickable (but nothing happens when you click on them). They should not have an active state, they should be static/flat.

mkllnk commented:
@daniellemoorhead I updated the PR and fixed all the easy stuff. Just a note on the *active* property labels: In production, these properties have descriptions and they pop up as a tooltip when you hover or tap on mobile. That's why they are active. But even without tooltip, they appeared active which I now fixed. I removed the darker border on hover completely but the pointer cursor is still displayed if there is a tooltip. I reckon that the tooltip is quite important to explain what chemical-free or organic really mean.

@yukoosawa Sorry to bug you with another two questions:

1. Are you okay with a one-line product description or does the additional line justify a complicated Javascript solution to display the `...` or a `more` button?
2. When products have only one variant, it often has the same name as the product. But it can differ. It usually differs when there are multiple variants. In your designs I have seen screens with variant name (above the unit info) and screens without the variant name (only unit info). Do you reckon we should come up with a clever algorithm when to display the variant name or should we always display it for consistency?

daniellemoorhead commented:
@mkllnk, thanks for your responses to the testing!

For # 1 comment
>I'm not sure how to resolve this. We show the product summary (image, name, description, supplier) and then we list all variants. Should I hide the variant name if it's the same as the product name? What if only one variant has the same name and the others are different? Treating them differently would look odd.

Solution as agreed by myself and @yukoosawa is to only display variant names if there is >1 variant name for the product. If there is only 1 variant, do not display that name. Hopefully this solves the problem 🤞 

For # 2 comment

>I know, that's because HTML can't display the ... after two lines of text. It only works for one. There are more complicated Javascript solutions but I was hoping to avoid those. How important is this second line?

Both Yuko and I agree that a single line description is not enough. But we'll leave it for now, Yuko will have a look at it when I stage this PR with all your padding changes and then we can make a decision then.

So, is this ready for testing again? Or code review?

daniellemoorhead commented:
Oh, and:

>@daniellemoorhead I updated the PR and fixed all the easy stuff. Just a note on the active property labels: In production, these properties have descriptions and they pop up as a tooltip when you hover or tap on mobile. That's why they are active. But even without tooltip, they appeared active which I now fixed. I removed the darker border on hover completely but the pointer cursor is still displayed if there is a tooltip. I reckon that the tooltip is quite important to explain what chemical-free or organic really mean.

This is a fine solution, we're pretty sure anecdotally that hardly anyone uses the description field, so this solution works fine 😄

mkllnk commented:
I fixed the variant name display as well now. This is ready for review again. The new commits start at "Remove margin of product listing on mobile and tablet".

daniellemoorhead commented:
Hey @mkllnk can this come back to test ready now? Luis said it's still looking good 6 days ago but there's been no movement since.

mkllnk commented:
Yes, I think one review is enough after my minor changes. It's test-ready.

daniellemoorhead commented:
Retested, images are now centre aligned, tested on samsung 8, iphone 8, ipad 6th gen.

Good to go 🎉

sigmundpetersen commented:
Hey @daniellemoorhead I wonder if the above comment was rather meant for #5263? Since that one was deployed earlier today and had the centre alignment testing remaining.

This PR is not what was staged recently at least, please double check 🙂

daniellemoorhead commented:
Oh god, you’re right, thanks Sigmund. Thankfully this one only merged to a branch so I can retest when the next story is added to that branch. Thanks for watching!

> On 2 Jun 2020, at 7:51 pm, Sigmund Petersen <notifications@github.com> wrote:
> 
> ﻿
> Hey @daniellemoorhead I wonder if the above comment was rather meant for #5263? Since that one was deployed earlier today and had the centre alignment testing remaining.
> 
> This PR is not what was staged recently at least, please double check 🙂
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
